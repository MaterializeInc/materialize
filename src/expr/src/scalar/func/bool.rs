// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Boolean functions.

use repr::strconv;
use repr::{Datum, RowArena, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::{EvalError, ScalarExpr};

pub const CAST_BOOL_TO_STRING_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::String),
};

pub fn cast_bool_to_string_explicit(a: Datum) -> Result<Datum, EvalError> {
    // N.B. this function differs from `cast_bool_to_string_implicit` because
    // the SQL specification requires `true` and `false` to be spelled out
    // in explicit casts, while PostgreSQL prefers its more concise `t` and `f`
    // representation in implicit casts.
    match a.unwrap_bool() {
        true => Ok(Datum::from("true")),
        false => Ok(Datum::from("false")),
    }
}

pub fn cast_bool_to_string_implicit(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::String(strconv::format_bool_static(a.unwrap_bool())))
}

pub const CAST_STRING_TO_BOOL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn cast_string_to_bool(a: Datum) -> Result<Datum, EvalError> {
    match strconv::parse_bool(a.unwrap_str())? {
        true => Ok(Datum::True),
        false => Ok(Datum::False),
    }
}

pub const AND_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn and<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::False => Ok(Datum::False),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::False) => Ok(Datum::False),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::True, Datum::True) => Ok(Datum::True),
            _ => unreachable!(),
        },
    }
}

pub const OR_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn or<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::True => Ok(Datum::True),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::True) => Ok(Datum::True),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::False, Datum::False) => Ok(Datum::False),
            _ => unreachable!(),
        },
    }
}

pub const NOT_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn not(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(!a.unwrap_bool()))
}
