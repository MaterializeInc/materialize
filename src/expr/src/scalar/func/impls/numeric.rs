// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use dec::{OrderedDecimal, Rounding};
use repr::adt::numeric::{self, Numeric};
use repr::strconv;

use crate::EvalError;

sqlfunc!(
    #[sqlname = "-"]
    #[preserves_uniqueness = true]
    fn neg_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().neg(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "abs"]
    fn abs_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().abs(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "ceilnumeric"]
    fn ceil_numeric(mut a: Numeric) -> Numeric {
        // ceil will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Ceiling);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "expnumeric"]
    fn exp_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.exp(&mut a);
        let cx_status = cx.status();
        if cx_status.overflow() {
            Err(EvalError::FloatOverflow)
        } else if cx_status.subnormal() {
            Err(EvalError::FloatUnderflow)
        } else {
            numeric::munge_numeric(&mut a).unwrap();
            Ok(a)
        }
    }
);

sqlfunc!(
    #[sqlname = "floornumeric"]
    fn floor_numeric(mut a: Numeric) -> Numeric {
        // floor will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Floor);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

fn log_guard_numeric(val: &Numeric, function_name: &str) -> Result<(), EvalError> {
    if val.is_negative() {
        return Err(EvalError::NegativeOutOfDomain(function_name.to_owned()));
    }
    if val.is_zero() {
        return Err(EvalError::ZeroOutOfDomain(function_name.to_owned()));
    }
    Ok(())
}

// From the `decNumber` library's documentation:
// > Inexact results will almost always be correctly rounded, but may be up to 1
// > ulp (unit in last place) in error in rare cases.
//
// See decNumberLog10 documentation at http://speleotrove.com/decimal/dnnumb.html
fn log_numeric<F>(mut a: Numeric, logic: F, name: &'static str) -> Result<Numeric, EvalError>
where
    F: Fn(&mut dec::Context<Numeric>, &mut Numeric),
{
    log_guard_numeric(&a, name)?;
    let mut cx = numeric::cx_datum();
    logic(&mut cx, &mut a);
    numeric::munge_numeric(&mut a).unwrap();
    Ok(a)
}

sqlfunc!(
    #[sqlname = "lnnumeric"]
    fn ln_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::ln, "ln")
    }
);

sqlfunc!(
    #[sqlname = "log10numeric"]
    fn log10_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::log10, "log10")
    }
);

sqlfunc!(
    #[sqlname = "roundnumeric"]
    fn round_numeric(mut a: Numeric) -> Numeric {
        // round will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        numeric::cx_datum().round(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "sqrtnumeric"]
    fn sqrt_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        if a.is_negative() {
            return Err(EvalError::NegSqrt);
        }
        let mut cx = numeric::cx_datum();
        cx.sqrt(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        Ok(a)
    }
);

sqlfunc!(
    #[sqlname = "numerictoi16"]
    fn cast_numeric_to_int16(mut a: Numeric) -> Result<i16, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        let i = cx.try_into_i32(a).or(Err(EvalError::Int16OutOfRange))?;
        i16::try_from(i).or(Err(EvalError::Int16OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi32"]
    fn cast_numeric_to_int32(mut a: Numeric) -> Result<i32, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i32(a).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi64"]
    fn cast_numeric_to_int64(mut a: Numeric) -> Result<i64, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i64(a).or(Err(EvalError::Int64OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictof32"]
    fn cast_numeric_to_float32(a: Numeric) -> Result<f32, EvalError> {
        let i = a.to_string().parse::<f32>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float32OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictof64"]
    fn cast_numeric_to_float64(a: Numeric) -> Result<f64, EvalError> {
        let i = a.to_string().parse::<f64>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float64OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictostr"]
    fn cast_numeric_to_string(a: Numeric) -> String {
        let mut buf = String::new();
        strconv::format_numeric(&mut buf, &OrderedDecimal(a));
        buf
    }
);
