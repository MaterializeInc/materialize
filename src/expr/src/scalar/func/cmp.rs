// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Comparison functions.

use repr::{Datum, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::EvalError;

pub const CMP_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a == b))
}

pub fn not_eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a != b))
}

pub fn lt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a < b))
}

pub fn lte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a <= b))
}

pub fn gt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a > b))
}

pub fn gte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a >= b))
}
