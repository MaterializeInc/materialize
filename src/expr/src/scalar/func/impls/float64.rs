// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use chrono::{DateTime, NaiveDateTime, Utc};
use mz_lowertest::MzReflect;
use mz_ore::cast::TryCastFrom;
use mz_repr::adt::numeric::{self, Numeric, NumericMaxScale};
use mz_repr::adt::timestamp::CheckedTimestamp;
use mz_repr::{strconv, ColumnType, ScalarType};
use serde::{Deserialize, Serialize};

use crate::scalar::func::EagerUnaryFunc;
use crate::scalar::DomainLimit;
use crate::EvalError;

sqlfunc!(
    #[sqlname = "-"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(NegFloat64)]
    fn neg_float64(a: f64) -> f64 {
        -a
    }
);

sqlfunc!(
    #[sqlname = "abs"]
    fn abs_float64(a: f64) -> f64 {
        a.abs()
    }
);

sqlfunc!(
    #[sqlname = "roundf64"]
    fn round_float64(a: f64) -> f64 {
        // f64::round violates IEEE 754 by rounding ties away from zero rather than
        // to nearest even. There appears to be no way to round ties to nearest even
        // in Rust natively, so bail out to C.
        extern "C" {
            fn rint(f: f64) -> f64;
        }
        unsafe { rint(a) }
    }
);

sqlfunc!(
    #[sqlname = "truncf64"]
    fn trunc_float64(a: f64) -> f64 {
        a.trunc()
    }
);

sqlfunc!(
    #[sqlname = "ceilf64"]
    fn ceil_float64(a: f64) -> f64 {
        a.ceil()
    }
);

sqlfunc!(
    #[sqlname = "floorf64"]
    fn floor_float64(a: f64) -> f64 {
        a.floor()
    }
);

sqlfunc!(
    #[sqlname = "double_to_smallint"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastInt16ToFloat64)]
    fn cast_float64_to_int16(a: f64) -> Result<i16, EvalError> {
        let f = round_float64(a);
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= (i16::MIN as f64)) && (f < -(i16::MIN as f64)) {
            Ok(f as i16)
        } else {
            Err(EvalError::Int16OutOfRange)
        }
    }
);

sqlfunc!(
    #[sqlname = "double_to_integer"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastInt32ToFloat64)]
    fn cast_float64_to_int32(a: f64) -> Result<i32, EvalError> {
        let f = round_float64(a);
        // This condition is delicate because i32::MIN can be represented exactly by
        // an f64 but not i32::MAX. We follow PostgreSQL's approach here.
        //
        // See: https://github.com/postgres/postgres/blob/ca3b37487/src/include/c.h#L1074-L1096
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= (i32::MIN as f64)) && (f < -(i32::MIN as f64)) {
            Ok(f as i32)
        } else {
            Err(EvalError::Int32OutOfRange)
        }
    }
);

sqlfunc!(
    #[sqlname = "f64toi64"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastInt64ToFloat64)]
    fn cast_float64_to_int64(a: f64) -> Result<i64, EvalError> {
        let f = round_float64(a);
        // This condition is delicate because i64::MIN can be represented exactly by
        // an f64 but not i64::MAX. We follow PostgreSQL's approach here.
        //
        // See: https://github.com/postgres/postgres/blob/ca3b37487/src/include/c.h#L1074-L1096
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= (i64::MIN as f64)) && (f < -(i64::MIN as f64)) {
            Ok(f as i64)
        } else {
            Err(EvalError::Int64OutOfRange)
        }
    }
);

sqlfunc!(
    #[sqlname = "double_to_real"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastFloat32ToFloat64)]
    fn cast_float64_to_float32(a: f64) -> Result<f32, EvalError> {
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        let result = a as f32;
        if result.is_infinite() && !a.is_infinite() {
            Err(EvalError::FloatOverflow)
        } else if result == 0.0 && a != 0.0 {
            Err(EvalError::FloatUnderflow)
        } else {
            Ok(result)
        }
    }
);

sqlfunc!(
    #[sqlname = "double_to_text"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastStringToFloat64)]
    fn cast_float64_to_string(a: f64) -> String {
        let mut s = String::new();
        strconv::format_float64(&mut s, a);
        s
    }
);

sqlfunc!(
    #[sqlname = "double_to_uint2"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastUint16ToFloat64)]
    fn cast_float64_to_uint16(a: f64) -> Result<u16, EvalError> {
        let f = round_float64(a);
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= 0.0) && (f <= (u16::MAX as f64)) {
            Ok(f as u16)
        } else {
            Err(EvalError::UInt16OutOfRange)
        }
    }
);

sqlfunc!(
    #[sqlname = "double_to_uint4"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastUint32ToFloat64)]
    fn cast_float64_to_uint32(a: f64) -> Result<u32, EvalError> {
        let f = round_float64(a);
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= 0.0) && (f <= (u32::MAX as f64)) {
            Ok(f as u32)
        } else {
            Err(EvalError::UInt32OutOfRange)
        }
    }
);

sqlfunc!(
    #[sqlname = "double_to_uint8"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastUint64ToFloat64)]
    fn cast_float64_to_uint64(a: f64) -> Result<u64, EvalError> {
        let f = round_float64(a);
        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        if (f >= 0.0) && (f <= (u64::MAX as f64)) {
            Ok(f as u64)
        } else {
            Err(EvalError::UInt64OutOfRange)
        }
    }
);

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct CastFloat64ToNumeric(pub Option<NumericMaxScale>);

impl<'a> EagerUnaryFunc<'a> for CastFloat64ToNumeric {
    type Input = f64;
    type Output = Result<Numeric, EvalError>;

    fn call(&self, a: f64) -> Result<Numeric, EvalError> {
        if a.is_infinite() {
            return Err(EvalError::InfinityOutOfDomain(
                "casting double precision to numeric".to_owned(),
            ));
        }
        let mut a = Numeric::from(a);
        if let Some(scale) = self.0 {
            if numeric::rescale(&mut a, scale.into_u8()).is_err() {
                return Err(EvalError::NumericFieldOverflow);
            }
        }
        match numeric::munge_numeric(&mut a) {
            Ok(_) => Ok(a),
            Err(_) => Err(EvalError::NumericFieldOverflow),
        }
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Numeric { max_scale: self.0 }.nullable(input.nullable)
    }

    fn inverse(&self) -> Option<crate::UnaryFunc> {
        to_unary!(super::CastNumericToFloat64)
    }
}

impl fmt::Display for CastFloat64ToNumeric {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("double_to_numeric")
    }
}

sqlfunc!(
    #[sqlname = "sqrtf64"]
    fn sqrt_float64(a: f64) -> Result<f64, EvalError> {
        if a < 0.0 {
            return Err(EvalError::NegSqrt);
        }
        Ok(a.sqrt())
    }
);

sqlfunc!(
    #[sqlname = "cbrtf64"]
    fn cbrt_float64(a: f64) -> f64 {
        a.cbrt()
    }
);

sqlfunc!(
    fn cos(a: f64) -> Result<f64, EvalError> {
        if a.is_infinite() {
            return Err(EvalError::InfinityOutOfDomain("cos".to_owned()));
        }
        Ok(a.cos())
    }
);

sqlfunc!(
    fn acos(a: f64) -> Result<f64, EvalError> {
        if a < -1.0 || 1.0 < a {
            return Err(EvalError::OutOfDomain(
                DomainLimit::Inclusive(-1),
                DomainLimit::Inclusive(1),
                "acos".to_owned(),
            ));
        }
        Ok(a.acos())
    }
);

sqlfunc!(
    fn cosh(a: f64) -> f64 {
        a.cosh()
    }
);

sqlfunc!(
    fn acosh(a: f64) -> Result<f64, EvalError> {
        if a < 1.0 {
            return Err(EvalError::OutOfDomain(
                DomainLimit::Inclusive(1),
                DomainLimit::None,
                "acosh".to_owned(),
            ));
        }
        Ok(a.acosh())
    }
);

sqlfunc!(
    fn sin(a: f64) -> Result<f64, EvalError> {
        if a.is_infinite() {
            return Err(EvalError::InfinityOutOfDomain("sin".to_owned()));
        }
        Ok(a.sin())
    }
);

sqlfunc!(
    fn asin(a: f64) -> Result<f64, EvalError> {
        if a < -1.0 || 1.0 < a {
            return Err(EvalError::OutOfDomain(
                DomainLimit::Inclusive(-1),
                DomainLimit::Inclusive(1),
                "asin".to_owned(),
            ));
        }
        Ok(a.asin())
    }
);

sqlfunc!(
    fn sinh(a: f64) -> f64 {
        a.sinh()
    }
);

sqlfunc!(
    fn asinh(a: f64) -> f64 {
        a.asinh()
    }
);

sqlfunc!(
    fn tan(a: f64) -> Result<f64, EvalError> {
        if a.is_infinite() {
            return Err(EvalError::InfinityOutOfDomain("tan".to_owned()));
        }
        Ok(a.tan())
    }
);

sqlfunc!(
    fn atan(a: f64) -> f64 {
        a.atan()
    }
);

sqlfunc!(
    fn tanh(a: f64) -> f64 {
        a.tanh()
    }
);

sqlfunc!(
    fn atanh(a: f64) -> Result<f64, EvalError> {
        if a < -1.0 || 1.0 < a {
            return Err(EvalError::OutOfDomain(
                DomainLimit::Inclusive(-1),
                DomainLimit::Inclusive(1),
                "atanh".to_owned(),
            ));
        }
        Ok(a.atanh())
    }
);

sqlfunc!(
    fn cot(a: f64) -> Result<f64, EvalError> {
        if a.is_infinite() {
            return Err(EvalError::InfinityOutOfDomain("cot".to_owned()));
        }
        Ok(1.0 / a.tan())
    }
);

sqlfunc!(
    fn radians(a: f64) -> f64 {
        a.to_radians()
    }
);

sqlfunc!(
    fn degrees(a: f64) -> f64 {
        a.to_degrees()
    }
);

sqlfunc!(
    #[sqlname = "log10f64"]
    fn log10(a: f64) -> Result<f64, EvalError> {
        if a.is_sign_negative() {
            return Err(EvalError::NegativeOutOfDomain("log10".to_owned()));
        }
        if a == 0.0 {
            return Err(EvalError::ZeroOutOfDomain("log10".to_owned()));
        }
        Ok(a.log10())
    }
);

sqlfunc!(
    #[sqlname = "lnf64"]
    fn ln(a: f64) -> Result<f64, EvalError> {
        if a.is_sign_negative() {
            return Err(EvalError::NegativeOutOfDomain("ln".to_owned()));
        }
        if a == 0.0 {
            return Err(EvalError::ZeroOutOfDomain("ln".to_owned()));
        }
        Ok(a.ln())
    }
);

sqlfunc!(
    #[sqlname = "expf64"]
    fn exp(a: f64) -> f64 {
        a.exp()
    }
);

sqlfunc!(
    #[sqlname = "mz_sleep"]
    fn sleep(a: f64) -> Option<CheckedTimestamp<DateTime<Utc>>> {
        let duration = std::time::Duration::from_secs_f64(a);
        std::thread::sleep(duration);
        None
    }
);

sqlfunc!(
    #[sqlname = "tots"]
    fn to_timestamp(f: f64) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
        const NANO_SECONDS_PER_SECOND: i64 = 1_000_000_000;
        if f.is_nan() {
            Err(EvalError::TimestampCannotBeNan)
        } else if f.is_infinite() {
            // TODO(jkosh44) implement infinite timestamps
            Err(EvalError::TimestampOutOfRange)
        } else {
            let mut secs = i64::try_cast_from(f.trunc()).ok_or(EvalError::TimestampOutOfRange)?;
            // NOTE(benesch): PostgreSQL has microsecond precision in its timestamps,
            // while chrono has nanosecond precision. While we normally accept
            // nanosecond precision, here we round to the nearest microsecond because
            // f64s lose quite a bit of accuracy in the nanosecond digits when dealing
            // with common Unix timestamp values (> 1 billion).
            let microsecs = (f.fract() * 1_000_000.0).round();
            let mut nanosecs =
                i64::try_cast_from(microsecs * 1_000.0).ok_or(EvalError::TimestampOutOfRange)?;
            if nanosecs < 0 {
                secs = secs.checked_sub(1).ok_or(EvalError::TimestampOutOfRange)?;
                nanosecs = NANO_SECONDS_PER_SECOND
                    .checked_add(nanosecs)
                    .ok_or(EvalError::TimestampOutOfRange)?;
            }
            // Ensure `nanosecs` is less than 1 second.
            secs = secs
                .checked_add(nanosecs / NANO_SECONDS_PER_SECOND)
                .ok_or(EvalError::TimestampOutOfRange)?;
            nanosecs %= NANO_SECONDS_PER_SECOND;
            let nanosecs = u32::try_from(nanosecs).map_err(|_| EvalError::TimestampOutOfRange)?;
            match NaiveDateTime::from_timestamp_opt(secs, nanosecs) {
                Some(ts) => {
                    let dt = DateTime::<Utc>::from_utc(ts, Utc);
                    CheckedTimestamp::from_timestamplike(dt)
                        .map_err(|_| EvalError::TimestampOutOfRange)
                }
                None => Err(EvalError::TimestampOutOfRange),
            }
        }
    }
);
