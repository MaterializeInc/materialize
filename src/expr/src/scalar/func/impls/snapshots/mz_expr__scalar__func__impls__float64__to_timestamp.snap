---
source: src/expr/src/scalar/func/impls/float64.rs
expression: "#[sqlfunc(\n    sqlname = \"tots\",\n    preserves_uniqueness = false,\n    inverse = None,\n    is_monotone = false,\n)]\n#[allow(clippy::extra_unused_lifetimes)]\npub fn to_timestamp<'a>(f: f64) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {\n    {\n        const NANO_SECONDS_PER_SECOND: i64 = 1_000_000_000;\n        if f.is_nan() {\n            Err(EvalError::TimestampCannotBeNan)\n        } else if f.is_infinite() {\n            Err(EvalError::TimestampOutOfRange)\n        } else {\n            let mut secs = i64::try_cast_from(f.trunc())\n                .ok_or(EvalError::TimestampOutOfRange)?;\n            let microsecs = (f.fract() * 1_000_000.0).round();\n            let mut nanosecs = i64::try_cast_from(microsecs * 1_000.0)\n                .ok_or(EvalError::TimestampOutOfRange)?;\n            if nanosecs < 0 {\n                secs = secs.checked_sub(1).ok_or(EvalError::TimestampOutOfRange)?;\n                nanosecs = NANO_SECONDS_PER_SECOND\n                    .checked_add(nanosecs)\n                    .ok_or(EvalError::TimestampOutOfRange)?;\n            }\n            secs = secs\n                .checked_add(nanosecs / NANO_SECONDS_PER_SECOND)\n                .ok_or(EvalError::TimestampOutOfRange)?;\n            nanosecs %= NANO_SECONDS_PER_SECOND;\n            let nanosecs = u32::try_from(nanosecs)\n                .map_err(|_| EvalError::TimestampOutOfRange)?;\n            match DateTime::from_timestamp(secs, nanosecs) {\n                Some(dt) => {\n                    CheckedTimestamp::from_timestamplike(dt)\n                        .map_err(|_| EvalError::TimestampOutOfRange)\n                }\n                None => Err(EvalError::TimestampOutOfRange),\n            }\n        }\n    }\n}\n"
---
#[derive(
    proptest_derive::Arbitrary,
    Ord,
    PartialOrd,
    Clone,
    Debug,
    Eq,
    PartialEq,
    serde::Serialize,
    serde::Deserialize,
    Hash,
    mz_lowertest::MzReflect
)]
pub struct ToTimestamp;
impl<'a> crate::func::EagerUnaryFunc<'a> for ToTimestamp {
    type Input = f64;
    type Output = Result<CheckedTimestamp<DateTime<Utc>>, EvalError>;
    fn call(&self, a: Self::Input) -> Self::Output {
        to_timestamp(a)
    }
    fn output_type(&self, input_type: mz_repr::ColumnType) -> mz_repr::ColumnType {
        use mz_repr::AsColumnType;
        let output = Self::Output::as_column_type();
        let propagates_nulls = crate::func::EagerUnaryFunc::propagates_nulls(self);
        let nullable = output.nullable;
        output.nullable(nullable || (propagates_nulls && input_type.nullable))
    }
    fn inverse(&self) -> Option<crate::UnaryFunc> {
        None
    }
    fn is_monotone(&self) -> bool {
        false
    }
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}
impl std::fmt::Display for ToTimestamp {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        f.write_str("tots")
    }
}
#[allow(clippy::extra_unused_lifetimes)]
pub fn to_timestamp<'a>(f: f64) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
    {
        const NANO_SECONDS_PER_SECOND: i64 = 1_000_000_000;
        if f.is_nan() {
            Err(EvalError::TimestampCannotBeNan)
        } else if f.is_infinite() {
            Err(EvalError::TimestampOutOfRange)
        } else {
            let mut secs = i64::try_cast_from(f.trunc())
                .ok_or(EvalError::TimestampOutOfRange)?;
            let microsecs = (f.fract() * 1_000_000.0).round();
            let mut nanosecs = i64::try_cast_from(microsecs * 1_000.0)
                .ok_or(EvalError::TimestampOutOfRange)?;
            if nanosecs < 0 {
                secs = secs.checked_sub(1).ok_or(EvalError::TimestampOutOfRange)?;
                nanosecs = NANO_SECONDS_PER_SECOND
                    .checked_add(nanosecs)
                    .ok_or(EvalError::TimestampOutOfRange)?;
            }
            secs = secs
                .checked_add(nanosecs / NANO_SECONDS_PER_SECOND)
                .ok_or(EvalError::TimestampOutOfRange)?;
            nanosecs %= NANO_SECONDS_PER_SECOND;
            let nanosecs = u32::try_from(nanosecs)
                .map_err(|_| EvalError::TimestampOutOfRange)?;
            match DateTime::from_timestamp(secs, nanosecs) {
                Some(dt) => {
                    CheckedTimestamp::from_timestamplike(dt)
                        .map_err(|_| EvalError::TimestampOutOfRange)
                }
                None => Err(EvalError::TimestampOutOfRange),
            }
        }
    }
}
